### PR TITLE
fix(acp): make /clear reach the agent and reject /compact it can't do

### LINKED ISSUE #10763: ACP: /clear and /compact don't reach the agent — context is only cleared goose-side, and the token counter lies
**Describe the bug**

`/clear` and `/compact` only mutate goose's own copy of the conversation. For ACP providers, where the *agent* holds the authoritative conversation, neither command reaches the agent — so the context the model actually sees is unchanged. The token counter is reset to zero at the same time, so the UI reports an empty session while the real window is still full.

The next prompt after a `/clear` can therefore fail immediately with `Prompt is too long`, with no way for the user to recover the session.

**To Reproduce**

1. Configure the `claude-acp` provider (Claude Code via `@agentclientprotocol/claude-agent-acp`).
2. Work in a session until the context is substantially full.
3. Run `/clear`. The token counter drops to 0 and the transcript empties.
4. Send one short prompt.
5. The turn fails with `Request failed: Internal error: Prompt is too long: {"errorKind": "invalid_request"}`.

**Expected behavior**

`/clear` should reset the agent-side conversation too (e.g. via `session/new`), so that the cleared state the user sees matches the state the model sees. `/compact` should either be applied agent-side or be rejected as unsupported for providers that manage their own context.

**Root cause analysis**

`handle_clear_command` in `crates/goose/src/agents/execute_commands.rs:184` does exactly two things:

```rust
manager.replace_conversation(session_id, &Conversation::default()).await?;
manager.update(session_id).usage(Usage::new(Some(0), Some(0), Some(0))).apply().await?;
Ok(Some(user_only_assistant_text("Conversation cleared")))
```

Both writes go to goose's session store. There is no provider call. `handle_compact_command` (line 152) is the same shape — `compact_messages` followed by `replace_conversation`, all local.

The ACP provider has no corresponding hook either; grepping `crates/goose/src/acp/provider.rs` for clear/reset/session-recreation turns up only a tool-update buffer and a line buffer.

This matters more for ACP than for a normal provider because `AcpProvider::stream()` sends **only the latest user message** — `messages_to_prompt(messages, false)` at `provider.rs:1365` returns just the trailing user content, deliberately relying on the agent to hold history. The agent's session *is* the context, so clearing goose's copy is cosmetic.

The zeroed usage counters make it silent: `check_if_compaction_needed` returns `Ok(false)` immediately when `provider.manages_own_context()` (`crates/goose/src/context_mgmt/mod.rs:221`, true for ACP), so nothing else recomputes occupancy either.

**Evidence**

From a diagnostics bundle for an affected session (v1.44.0, `claude-acp`):

```
15:19:14  Restoring evicted session <id>
15:23:05  msg 0: user      "/clear"
15:23:05  msg 1: assistant "Conversation cleared"
15:23:21  msg 2: user      <one short prompt>
15:23:26  msgs 3-20: tool calls
15:24:42  ERROR goose::agents::agent :: Request failed: Internal error: Prompt is too long
15:24:42  WARN  goose::agents::agent :: Failed to generate session description: ... Prompt is too long
```

The stored conversation at failure time is 23 messages / 27,766 chars — roughly 7k tokens — and `session.usage` reads all zeros. The pre-`/clear` history, which is what actually filled the window, is still resident in the agent.

**Please provide the following information**
- **OS & Arch:** macOS (Apple Silicon)
- **Interface:** UI (Desktop)
- **Version:** v1.44.0
- **Provider & Model:** Claude Code (`claude-acp` via `@agentclientprotocol/claude-agent-acp`)

**Additional context**

Related but distinct: the uncapped conversation replay on ACP session restore (filed separately) — together these two make a long ACP session unrecoverable, since neither clearing nor restoring reduces what gets sent.

### PR DESCRIPTION
**Pull request by kojiromike**
Fixes #10763.

## Problem

For providers that manage their own context, goose's transcript is a display copy — `AcpProvider::stream()` sends only the trailing user message and relies on the agent to hold history. `/clear` replaced that copy with an empty conversation and zeroed the token counters without telling the provider anything, so the model kept the full window while the UI reported an empty session. The next prompt could fail with `Prompt is too long` and the session was unrecoverable.

`/compact` had the same shape: it summarized a conversation the model never sees, spending a summarization call to no effect.

## Change

`Provider::reset_context()`, whose default is derived from `manages_own_context()` rather than being unconditionally `Ok`. A provider holding the authoritative history has to *opt in* to claiming it can drop it — so a context-owning provider added later can't inherit a silent success and quietly reintroduce this bug. That addresses @reuvenneumann's point that the hook should hang off `manages_own_context()` rather than any provider-specific path.

- `AcpProvider` implements it: opens a fresh `session/new`, swaps it in, closes the superseded session, and clears the handoff-context latch, the agent-reported context size, and the applied-model cache so the new session is configured from scratch. This covers **every** `*-acp` provider — `claude-acp`, `codex-acp`, `amp-acp`, `copilot-acp` and `pi-acp` all resolve to `type Provider = AcpProvider` — which matches the independent `codex-acp` reproduction in the issue thread.
- Both `/clear` entry points call it: the agent command and the CLI's separate `handle_clear`, which as noted in the thread the agent path never reaches.
- The current goose mode is re-sent to the new session for agents that expose mode as a config option rather than `session.modes`, which `session/new` alone does not carry over. Best-effort: the reset has already succeeded at that point, so a failure warns rather than reporting a reset that did happen as failed.
- The client loop tracks the active ACP session id and drops `session/update` notifications for a superseded session, so a trailing usage or mode update can't overwrite the freshly reset state.
- When a provider owns its context but cannot reset it (the Claude Code and Gemini CLI wrappers), `/clear` says so and **leaves the token counters alone**, so the meter keeps reflecting what the model can still see rather than reporting zero.
- `/compact` is rejected wherever `manages_own_context()` is true, in both the agent command and the CLI (which skips its confirmation prompt rather than asking and then refusing).

The Desktop path needs no change: its `/clear` forwards to the backend, which routes through `handle_clear_command` and emits `HistoryReplaced`.

## Tests

`crates/goose/tests/clear_command.rs`:
- `/clear` resets provider context and zeroes usage
- `/clear` preserves usage and warns when the provider cannot reset
- a context-owning provider that never overrides `reset_context` does **not** inherit a silent success (guards the default)
- `/compact` is rejected and leaves the transcript untouched

`acp/provider.rs`:
- `reset_context` opens a fresh session, closes the old one, and resets the latches
- a refused `session/new` leaves the current session in place and does not close it
- `reset_context` re-sends the mode to a config-option agent, and sends nothing to an agent without a mode config option
- updates for a superseded session id are ignored; updates predating any session are still accepted

`cargo clippy --all-targets -D warnings` clean. Pre-existing failures on `main` (`gcpauth` ×3, `chatgpt_codex` JWT, `prompt_manager` snapshot) are unchanged.

## Not addressed here

- The Claude Code and Gemini CLI wrappers now report honestly instead of resetting. Both are fixable — Gemini CLI just needs its `cli_session_id` `OnceLock` to become resettable, Claude Code needs its persistent process recycled — but neither is on the reported path and I can't exercise them end-to-end. Happy to file or fold in.
- @reuvenneumann's adjacent observation that `GOOSE_CONTEXT_LIMIT` only affects the display until the agent reports a `context_size` is a separate concern and unchanged by this PR.

### CODE REVIEW

File: crates/goose/src/acp/provider.rs [hunk 1]
```diff
@@ -675,9 +674,60 @@ impl Provider for AcpProvider {
...
+            ProviderError::ExecutionError(format!(
+                "ACP {} failed: {e}",
+                AGENT_METHOD_NAMES.session_new
+            ))
+        })?;
+
+        let previous = std::mem::replace(
+            &mut *self.session.lock().expect("ACP session lock poisoned"),
+            session,
+        );
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply crates/goose/src/acp/provider.rs:697]: Fixed in 7ed2ef001 — this was a real gap. `apply_session_mode` runs on the `session/new` reply, but it only acts through the `session.modes` branch, so an agent that advertises mode as a config option came back in the agent default while goose still reported the previously selected mode. `apply_session_config_options` doesn't cover it either: that replays the *statically configured* options, not a mode the user selected at runtime via `/mode`.

`reapply_mode_config_option()` now re-sends the current goose mode after the swap, gated on `session_has_config_option(Mode)` so the `session.modes` path isn't double-applied.

One deliberate departure from your wording, which said to reapply "before returning success." I made it best-effort — warn, don't fail. By the time it runs, the session swap has already happened and the context genuinely is cleared. Returning `Err` would make `/clear` report failure, and per this PR's own design the caller then *preserves the token counters* — so the meter would show a full window that no longer exists. That's a worse lie than the one this PR set out to fix, and it isn't recoverable by retrying. A wrong mode is recoverable with `/mode`.

Test: `reset_context_reapplies_mode_to_a_config_option_agent`, plus an assertion on the existing reset test that an agent *without* a mode config option isn't sent a spurious one.

File: crates/goose/src/acp/provider.rs [hunk 2]
```diff
@@ -675,9 +674,60 @@ impl Provider for AcpProvider {
...
+        })?;
+
+        let session = open_session(tx).await.map_err(|e| {
+            ProviderError::ExecutionError(format!(
+                "ACP {} failed: {e}",
+                AGENT_METHOD_NAMES.session_new
+            ))
+        })?;
+
+        let previous = std::mem::replace(
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply crates/goose/src/acp/provider.rs:694]: Fixed in 7ed2ef001. The client loop now holds an `active_session` id, set on every successful `session/new` reply, and the notification handler drops updates whose `session_id` doesn't match before touching `context_size`, mode, or the desktop notification callback.

Updates arriving before any session exists can't be attributed to a superseded session, so those are still accepted — that's the pre-first-session case, not a stale one. Lock poisoning also fails open, since dropping real updates is worse than accepting a stale one.

Worth noting the latch is complete on this branch because `NewSession` is the only session-creating request here; there's no `session/load` path on `main` to also cover.

Test: `updates_for_a_superseded_session_are_ignored`.

File: documentation/docs/guides/goose-cli-commands.md [hunk 3]
```diff
@@ -808,7 +808,7 @@ Once you're in an interactive session (via `goose session` or `goose run --inter
 **Available Commands:**
 - **`/?` or `/help`** - Display the help menu
 - **`/builtin <names>`** - Add builtin extensions by name (comma-separated)
-- **`/clear`** - Clear the current chat history
+- **`/clear`** - Clear the current chat history. With providers that keep their own conversation (ACP agents, Claude Code, Gemini CLI) this also starts the agent-side conversation over, so the model forgets what you just cleared
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply documentation/docs/guides/goose-cli-commands.md]: Fixed in 7ed2ef001 — you're right, the sentence promised exactly what the code refuses to do. Both wrappers return `ProviderError::NotImplemented` and `/clear` warns that the previous conversation is still resident.

Reworded to scope the agent-side reset to ACP agents, and to describe what Claude Code and Gemini CLI actually do: empty goose's transcript, warn that the model still holds the prior conversation, and leave the token counter reflecting what the model can still see.

File: crates/goose/src/agents/execute_commands.rs [hunk 4]
```diff
@@ -507,6 +524,23 @@ impl Agent {
...
+    format!(
+        "Cleared goose's transcript, but '{provider_name}' still holds the previous conversation \
+         and will keep sending it to the model: {error}. Start a new session for a full reset."
+    )
+}
+
+pub fn compaction_unsupported_message(provider_name: &str) -> String {
+    format!(
+        "Compaction is not supported with '{provider_name}' because it manages its own \
+         conversation context. Use /clear to start the conversation over."
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply crates/goose/src/agents/execute_commands.rs]: Fixed in eb9392e23 — the advice was self-defeating for exactly the two providers that need it most.

Reworded rather than adding a `can_reset_context()` predicate, which was my first instinct. Reset support is only truly knowable by attempting the reset, so a separate flag would be a second source of truth that can drift from what `reset_context()` actually does — the same class of problem the derived default was added to prevent. A provider could advertise `true` and still fail, and the message would lie again.

Instead the message defers to `/clear`'s own report, which already distinguishes the two cases: it either clears the agent-side conversation or says the provider still holds it. The escalation path is now self-revealing:

> Compaction is not supported with '{provider}' because it manages its own conversation context. Try /clear, which starts the agent's conversation over for providers that can reset it; if it reports that the previous conversation is still held, start a new session instead.

The `/compact` entry in `goose-cli-commands.md` carried the same unqualified "use `/clear` instead" and is updated to match.

File: crates/goose/src/acp/provider.rs [hunk 5]
```diff
@@ -1102,6 +1212,9 @@ async fn handle_requests(
                 let result = match session {
                     Ok(session) => {
                         session_ids.push(session.session_id.clone());
+                        if let Ok(mut guard) = active_session.lock() {
+                            *guard = Some(session.session_id.clone());
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply crates/goose/src/acp/provider.rs]: Fixed in 80b588694 — you are right that the latch was set too early. The id is now recorded only when the `NewSession` arm produces an `Ok` result, i.e. after both `apply_session_config_options` and `apply_session_mode` have succeeded, so a session that fails setup never becomes the filter's active session. That keeps the latch consistent with what `open_session` actually hands back: if setup fails, `reset_context` gets an `Err`, `self.session` stays on the old id, and updates for that old session keep being accepted.

Rolling back on error would have been the other option, but there is nothing correct to roll back *to* in general — the previous id lives in the provider, not the client loop. Only setting it on success needs no such knowledge.

One correction to the failure mode you described: the `?` on `apply_session_config_options` predates this PR (it is on the base commit, `7f634e536~1`), so a config-option rejection tears down the whole request loop rather than leaving it serving the old session. I left that alone as out of scope for this PR, but it is worth its own issue — a rejected saved config option should be reported to the `NewSession` caller, not kill the connection. Say the word and I will file it.

No unit test for this one: the `NewSession` arm lives inside `handle_requests`, which needs a real `ConnectionTo<Agent>`, and the existing `reset_context` tests mock at the `ClientRequest` channel boundary above it. `cargo test -p goose --lib acp::provider` (58) and `--test clear_command` (4) still pass, clippy clean.

File: crates/goose/src/agents/execute_commands.rs [hunk 6]
```diff
@@ -184,11 +191,21 @@ impl Agent {
     async fn handle_clear_command(&self, session_id: &str) -> Result<Option<Message>> {
         use crate::conversation::Conversation;
 
+        let provider = self.provider().await?;
+        let provider_reset = provider.reset_context().await;
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply crates/goose/src/agents/execute_commands.rs]: Fixed in 891735a2e — confirmed reachable, and confirmed ours. `handle_clear_command` did not touch the provider before this branch, so the `?` I added is the only thing that can produce `Provider not set` here.

On reachability: `restore_provider_from_session` failing is only `warn!`-logged, and the default-provider fallback right below it is gated on `if let Some(provider) = &*self.default_provider.read().await`. With no default installed, `get` returns an `Agent` whose `provider()` is `Err`, exactly as you describe.

Took the local-only-clear reading rather than reordering: a missing provider means there is no provider-side context that can go stale, so it is not the same situation as a provider that refused. `reset_context()` is now attempted only when a provider exists, and the stale-context warning only fires when one actually refused — reordering alone would have cleared locally but still surfaced a "the provider still holds your conversation" message naming a provider that is not there.

Regression test added in `crates/goose/tests/clear_command.rs` (`clear_falls_back_to_a_local_clear_without_a_provider`). Verified it fails with `Error: Provider not set` against the previous commit and passes now.

File: crates/goose/src/acp/provider.rs [hunk 7]
```diff
@@ -675,9 +700,70 @@ impl Provider for AcpProvider {
...
+        let (_, available) = resolve_model_info(&self.name, &self.session().response)?;
         Ok(available)
     }
+
+    async fn reset_context(&self) -> Result<(), ProviderError> {
+        let tx = self.tx.as_ref().ok_or_else(|| {
+            ProviderError::ExecutionError("ACP client is unavailable".to_string())
+        })?;
+
+        let session = open_session(tx).await.map_err(|e| {
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply crates/goose/src/acp/provider.rs:687]: Fixed in cfadb896e. You made the case I was missing: I called this out-of-scope on the last round because the `?` predates the branch, but this PR is what makes it reachable. `reset_context` calls `open_session` on every `/clear`, so a replacement session that refuses a saved config option now tears down the loop mid-session — where before it could only happen at initial connect, and the failure mode you describe (canceled oneshot, `self.session` still on the old id, loop and child gone, next prompt dead with "client unavailable") is worse than an error return. Latent-but-unreachable became reachable on this branch's own path, so it belongs here.

The `NewSession` arm now matches on `apply_session_config_options` and folds the error into `result`, so it goes out through `response_tx` exactly like the adjacent `apply_session_mode` failure already did. `open_session` gets a real `Err`, the loop keeps serving, and the active-session latch from 80b588694 still only fires on `Ok` — so a config-option rejection leaves the provider on the old session with its updates still accepted, rather than orphaning it.

Still no unit test, for the same reason as 80b588694: the arm lives inside `handle_requests`, which needs a real `ConnectionTo<Agent>`, while the `reset_context` tests mock at the `ClientRequest` channel boundary above it. Reaching it would mean driving the full in-process ACP server fixture with a deliberately bogus `session_config_options` entry, which needs a new field on the shared `TestConnectionConfig`. Happy to build that if you think this path warrants the fixture surface.

`cargo clippy -p goose --all-targets` clean, `--test clear_command` (5) passes.

#10775 is now redundant — closing it.

File: crates/goose/src/agents/execute_commands.rs [hunk 8]
```diff
@@ -184,11 +191,24 @@ impl Agent {
     async fn handle_clear_command(&self, session_id: &str) -> Result<Option<Message>> {
         use crate::conversation::Conversation;
 
+        let provider = self.provider().await.ok();
+        let provider_reset = match &provider {
+            Some(provider) => provider.reset_context().await,
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply crates/goose/src/agents/execute_commands.rs]: Fixed in 9a47557a1 — the local clear now runs first, and the provider is only reset after `replace_conversation` succeeds, so a failed local clear leaves both histories in place. The missing-provider fallback and the kept-counters-on-refusal behavior are unchanged.

Test: `clear_empties_the_local_transcript_before_resetting_the_provider`, via a `ResetOrderProbe` provider that records whether goose's transcript was already empty when `reset_context` arrived. Direct failure injection wasn't reachable through the public `SessionManager` surface — `replace_conversation` is a no-op for a missing session — so the test pins the ordering itself rather than the failure path.

File: crates/goose-cli/src/session/mod.rs [hunk 9]
```diff
@@ -1039,6 +1042,9 @@ impl CliSession {
     }
 
     async fn handle_clear(&mut self) -> Result<()> {
+        let provider = self.agent.provider().await?;
+        let provider_reset = provider.reset_context().await;
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply crates/goose-cli/src/session/mod.rs]: Fixed in 9a47557a1, same commit as the equivalent reorder in `handle_clear_command` — the CLI now runs `replace_conversation` first and only calls `reset_context` after it succeeds. `self.messages` is also cleared immediately after the persisted clear succeeds (rather than separately on each exit path), so the in-memory transcript can no longer outlive the persisted one and be handed to a fresh ACP session via the handoff path.

File: crates/goose/src/acp/provider.rs [hunk 10]
```diff
@@ -384,13 +383,39 @@ impl AcpProvider {
...
+        };
+        let Some(mode_id) = select_mode_id(candidates, self.session().response.modes.as_ref())
+        else {
+            return;
+        };
+        if let Err(e) = self
+            .send_set_config_option("", "mode".into(), mode_id)
+            .await
+        {
+            tracing::warn!(error = %e, "failed to reapply mode to the reset ACP session");
```
[the remark this answers is not in the document]
Comment (kojiromike) [reply crates/goose/src/acp/provider.rs]: Fixed in bcee70bf1 — and you've dissolved the tension I defended last round. The warn-don't-fail choice was only forced because the reapply ran *after* the swap, where a hard failure would misreport a context that genuinely had been cleared. The real problem was the placement.

The config-option mode application now travels with the rest of session setup in the client loop's `session/new` arm, next to `apply_session_config_options` and the existing `session.modes` handling. A refusal fails the whole `session/new` through `response_tx`, so `reset_context` gets an `Err` *before* any swap: the old session stays in place and latched, `/clear` reports failure, and the "still holds the previous conversation" message is accurate because it does. No success report can now coexist with an unknown agent-side permission mode. Agents exposing both surfaces get the config option, matching `update_mode`'s preference.

`reapply_mode_config_option` is deleted; the decision logic is extracted into `initial_session_mode`, a pure function unit-tested over both surfaces (config-option selected, config-option preferred over `session.modes`, neither surface, set_mode only-when-different, and candidate-not-offered failing setup).

**Comment by DOsinga** [comment]
fixed in https://github.com/aaif-goose/goose/pull/11094 as an alternative. thanks for your efforts, @kojiromike
